code/datums/components/bump_attack.dm

```diff
@@ -0,0 +1,51 @@
+/datum/component/bump_attack
+    var/active = FALSE
+    var/action/bump_attack_toggle/toggle_action
+
+/datum/component/bump_attack/Initialize()
+    . = ..()
+    if(!ismovableatom(parent))
+         return COMPONENT_INCOMPATIBLE
+    if(ismob(parent))
+        toggle_action = new()
+        RegisterSignal(parent, COMSIG_ACTION_TRIGGER, .proc/bump_attack_toggle)
+    else
+        return COMPONENT_INCOMPATIBLE
+
+/datum/component/bump_attack/proc/bump_attack_toggle(datum/source)
+    if(isliving(parent))
+	    var/mob/living/bumper = parent
+    else
+        return
+
+	active = !active
+	to_chat(bumper, "<span class='notice'>You will now [active ? "attack" : "push"] enemies who are in your way.</span>")
+
+	if(active)
+		bumper.RegisterSignal(xeno_bumper, COMSIG_MOVABLE_BUMP, .proc/bump_attack)
+	else
+		bumper.UnregisterSignal(xeno_bumper, COMSIG_MOVABLE_BUMP)
+        toggle_action.update_button_icon(active)
+
+/mob/living/proc/bump_attack(datum/source, atom/target)
+    if(!isliving(target))
+        return//we still want to push structures away and open doors
+	if(throwing || incapacitated() || next_move > world.time || a_intent == INTENT_HELP || a_intent == INTENT_GRAB)
+		return COMPONENT_MOVABLE_BUMP_RESOLVED
```

Comment:
Do we want to return this value here, instead of doing a bump?
If we are `throwing` we want to crash, for example, and if we are on `INTENT_HELP` we want to swap positions.

Reply:
hm. i'll change this a bit